src/Orleans/ApplicationParts/AssemblyAttributeFeatureProvider.cs

```diff
@@ -0,0 +1,34 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Orleans.ApplicationParts
+{
+    public sealed class AssemblyAttributeFeatureProvider<TFeature> : IApplicationFeatureProvider<TFeature>
+    {
+        public bool Equals(AssemblyAttributeFeatureProvider<TFeature> other) => true;
+
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            return obj.GetType() == this.GetType();
+        }
+
+        public override int GetHashCode()
+        {
+            return typeof(AssemblyAttributeFeatureProvider<TFeature>).GetHashCode();
+        }
+
+        public void PopulateFeature(IEnumerable<ApplicationPart> parts, TFeature feature)
+        {
+            foreach (var part in parts.OfType<AssemblyPart>())
+            {
+                var attributes = part.Assembly.GetCustomAttributes(true);
+                foreach (var attribute in attributes.OfType<IFeaturePopulator<TFeature>>())
```

Comment:
I don't follow this logic. The assembly attribute itself implements `IFeaturePopulator<T>`?

Reply:
yes, that's correct

Reply:
So we have an assembly level attribute like this:
```C#
[assembly: OrleansGeneratedCode.OrleansCodeGenb8790f1e97FeaturePopulator()]
```

Reply:
Yeah, I downloaded the code and saw that. It does work, although it's a little odd to have the populators be defined as attributes themselves, instead of just POCOs that can easily be used outside of the assembly populator (granted, it *can* be used, but it's odd that it is an attribute).
Maybe it's more appropriate to have a normal POCO and then have a well known assembly level attributes defined that points to it?

```
[assembly: FeaturePopulator(typeof(OrleansGeneratedCode.OrleansCodeGenb8790f1e97FeaturePopulator))]
```

Reply:
How could it be used outside of the assembly populator given it has a randomly generated name? Users can follow the same logic with the `attributes.OfType<IFeaturePopulator<T>>`. I'm not seeing the downside of this approach but I can change it if you prefer. It just means we need to use `Activator.CreateInstance()` to create it

Reply:
That actually I found a little odd. What's the reason for not using the AssemblyName as the differentiator instead of a random string?

Reply:
Can do that. No one should be referencing generated code directly, though

Reply:
Back in the olden days, when per-interface GrainFactory instances were being generated, that was normal, but that was years ago.

Reply:
Oh, the reason for not using AssemblyName was that I wasn't sure what characters were valid in AssemblyName (eg whitespace), so I would have to sanitize it into what's available in a type name. Anyway, I'll convert to some well-known attribute with the randomly generated name for now. Can't use a fixed name because `[InternalsVisibleTo]` might make it ambiguous